dao-voting/src/main.sw

```diff
@@ -0,0 +1,311 @@
+contract;
+
+dep abi;
+dep data_structures;
+dep errors;
+
+use std::{
+    address::Address,
+    assert::require,
+    block::height,
+    chain::auth::{AuthError, msg_sender},
+    context::{call_frames::msg_asset_id, msg_amount, this_balance},
+    contract_id::ContractId,
+    identity::Identity,
+    result::*,
+    revert::revert,
+    storage::StorageMap,
+    token::{transfer_to_output, force_transfer_to_contract},
+};
+
+use abi::DaoVoting;
+use data_structures::{CallData, Proposal};
+use errors::Error;
+
+storage {
+    gov_token: ContractId,
+    voting_period: u64,
+    approval_percentage: u64,
+    proposals: StorageMap<u64,
+    Proposal>, proposal_count: u64,
+    // The amount of governance tokens a user has deposited
+    balances: StorageMap<Identity,
+    u64>,
+    // The amount of votes a user has used on a proposal
+    votes: StorageMap<(Identity, u64),
+    u64>, state: u64, 
+}
+
+impl DaoVoting for Contract {
+    /// Initialize the dao with the governance token, voting parameters, and the proposal.
+    ///
+    /// # Parameters
+    ///
+    /// gov_token - contract id of the token to use to vote on governance proposals
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor is called more than once
+    /// - The voting period is not greater than 0
+    /// - The approval percentage is not greater than 0
+    #[storage(read, write)]
+    fn constructor(gov_token: ContractId) -> bool {
+        require(storage.state == 0, Error::CannotReinitialize);
+
+        storage.gov_token = gov_token;
+        storage.proposal_count = 0;
```

Comment:
The compiler can optimize this away, so it shouldn't hurt to leave it in from a gas cost perspective. We should avoid gas optimizations at this state and trust the compiler to do everything properly eventually.

Reply:
Fair point for optimization however for code quality I will argue that this is a bad practice since it is not needed and it adds additional cognitive load to process the info as to why it is there.